contracts/neutron-interchain-txs/src/contract.rs

```diff
@@ -0,0 +1,472 @@
+        QueryMsg::IcaContract {
+            interchain_account_id,
+        } => query_interchain_address_contract(deps, env, interchain_account_id),
+    }
+}
+
+pub fn query_interchain_address(
+    deps: Deps<InterchainQueries>,
+    env: Env,
+    interchain_account_id: String,
+    connection_id: String,
+) -> ContractResult<Binary> {
+    let query = InterchainQueries::InterchainAccountAddress {
+        owner_address: env.contract.address.to_string(),
+        interchain_account_id,
+        connection_id,
+    };
+
+    let res: QueryInterchainAccountAddressResponse = deps.querier.query(&query.into())?;
+    Ok(to_binary(&res)?)
+}
+
+pub fn query_interchain_address_contract(
+    deps: Deps<InterchainQueries>,
+    env: Env,
+    interchain_account_id: String,
+) -> ContractResult<Binary> {
+    let key =
+        "icacontroller-".to_string() + env.contract.address.as_str() + "." + &interchain_account_id;
+    let address = INTERCHAIN_ACCOUNTS.load(deps.storage, key)?;
+
+    Ok(to_binary(&address)?)
+}
+
+pub fn get_next_id(store: &mut dyn Storage) -> StdResult<u64> {
+    REPLY_QUEUE_ID
+        .keys(store, None, None, cosmwasm_std::Order::Descending)
+        .next()
+        .unwrap_or(Ok(IBC_SUDO_ID_RANGE_START))
+        .map(|id| id + 1)
+}
+
+pub fn save_reply_payload(store: &mut dyn Storage, payload: SudoPayload) -> StdResult<u64> {
+    let id = get_next_id(store)?;
+    REPLY_QUEUE_ID.save(store, id, &to_vec(&payload)?)?;
+    Ok(id)
+}
+
+pub fn read_reply_payload(store: &mut dyn Storage, id: u64) -> StdResult<SudoPayload> {
+    let data = REPLY_QUEUE_ID.load(store, id)?;
+    from_binary(&Binary(data))
+}
+
+pub fn read_sudo_payload(store: &mut dyn Storage, seq_id: u64) -> StdResult<SudoPayload> {
+    let data = SUDO_PAYLOAD.load(store, seq_id)?;
+    from_binary(&Binary(data))
+}
+
+fn msg_with_sudo_callback<C: Into<CosmosMsg<T>>, T>(
+    deps: DepsMut,
+    msg: C,
+    payload: SudoPayload,
+) -> StdResult<SubMsg<T>> {
+    let id = save_reply_payload(deps.storage, payload)?;
+    Ok(SubMsg::reply_on_success(msg, id))
+}
+
+fn execute_register_ica(
+    deps: DepsMut,
+    env: Env,
+    connection_id: String,
+    interchain_account_id: String,
+) -> StdResult<Response<NeutronMsg>> {
+    let register =
+        NeutronMsg::register_interchain_account(connection_id, interchain_account_id.clone());
+    let key =
+        "icacontroller-".to_string() + env.contract.address.as_str() + "." + &interchain_account_id;
+    INTERCHAIN_ACCOUNTS.save(deps.storage, key, &None)?;
+    Ok(Response::new().add_message(register))
+}
+
+fn execute_delegate(
+    mut deps: DepsMut,
+    env: Env,
+    interchain_account_id: String,
+    validator: String,
+    amount: u128,
+) -> StdResult<Response<NeutronMsg>> {
+    let key =
+        "icacontroller-".to_string() + env.contract.address.as_str() + "." + &interchain_account_id;
+
+    let (delegator, connection_id) = INTERCHAIN_ACCOUNTS
```

Comment:
i think that we should call this `interchain_acc_address` to be as explicit as possible